mailpoet/lib/Config/Populator.php

```diff
@@ -509,134 +505,116 @@ protected function newsletterOptionFields() {
+    $connection = $this->entityManager->getConnection();
+    $existingOptionFields = $connection->createQueryBuilder()
+      ->select('of.name, of.newsletter_type')
+      ->from($tableName, 'of')
+      ->executeQuery()
+      ->fetchAllAssociative();
+
+    // 2. Insert new option fields using a single query (good for first installs).
+    $inserts = array_udiff(
+      $optionFields,
+      $existingOptionFields,
+      fn($a, $b) => [$a['name'], $a['newsletter_type']] <=> [$b['name'], $b['newsletter_type']]
+    );
+    if ($inserts) {
+      $placeholders = implode(',', array_fill(0, count($inserts), '(?, ?)'));
+      $connection->executeStatement(
+        "INSERT INTO $tableName (name, newsletter_type) VALUES $placeholders",
+        array_merge(
+          ...array_map(
+            fn($of) => [$of['name'], $of['newsletter_type']],
+            $inserts
+          )
+        )
+      );
+    }
   }
 
-  protected function newsletterTemplates() {
+  private function populateNewsletterTemplates(): void {
+    // 1. Load templates from the file system.
     $templates = [];
     foreach ($this->templates as $template) {
       $template = self::TEMPLATES_NAMESPACE . $template;
       $template = new $template(Env::$assetsUrl);
       $templates[] = $template->get();
     }
-    return [
-      'rows' => $templates,
-      'identification_columns' => [
-        'name',
-      ],
-      'remove_duplicates' => true,
-    ];
-  }
-
-  protected function populate($model) {
-    $modelMethod = Helpers::underscoreToCamelCase($model);
-    $table = $this->prefix . $model;
-    $dataDescriptor = $this->$modelMethod();
-    $rows = $dataDescriptor['rows'];
-    $identificationColumns = array_fill_keys(
-      $dataDescriptor['identification_columns'],
-      ''
-    );
-    $removeDuplicates =
-      isset($dataDescriptor['remove_duplicates']) && $dataDescriptor['remove_duplicates'];
 
-    foreach ($rows as $row) {
-      $existenceComparisonFields = array_intersect_key(
-        $row,
-        $identificationColumns
-      );
+    // 2. Load existing corresponding (readonly) templates from the database.
+    $tableName = $this->entityManager->getClassMetadata(NewsletterTemplateEntity::class)->getTableName();
+    $connection = $this->entityManager->getConnection();
+    $existingTemplates = $connection->createQueryBuilder()
+      ->select('t.name, t.categories, t.readonly, t.thumbnail, t.body')
+      ->from($tableName, 't')
+      ->where('t.readonly = 1')
+      ->executeQuery()
+      ->fetchAllAssociativeIndexed();
+
+    // 3. Compare the existing and file system templates.
+    $inserts = [];
+    $updates = [];
+    foreach ($templates as $template) {
+      $existing = $existingTemplates[$template['name']] ?? null;
+      if (
+        $existing
+        && $existing['categories'] === $template['categories']
+        && $existing['body'] === $template['body']
+        && $existing['thumbnail'] === $template['thumbnail']
+      ) {
```

Comment:
There might be other considerations that I'm not aware of, but I'd rather compare the hash of the entire templates with `hash('sha256', serialize($var))`.

Still, this LGTM as is!

Reply:
@mhadidg That's a cool idea, I don't really like the long IF condition.

I'm looking into it, but there are also some issues:
1. It would also include the value of `readonly`. The PHP templates return it as in integer `1`, while from the DB it comes as a string `'1'`, so I would need to normalize it.
2. The row from the database and the array from PHP would both need to be sorted the same way.
3. It probably has a bit higher CPU cost, but I suppose it's maybe not even measurable.

So in the end, I'm more like let's maybe leave it as it is, it's just 3 compares, after all.

Reply:
Oh, and it's important to say that all of the three checked values are just strings.

Reply:
Well, it doesn't seem like a good idea then. Thanks for looking into it!